--- old/src/hotspot/share/gc/g1/g1CollectedHeap.cpp	2020-09-04 13:56:58.177010048 +0200
+++ new/src/hotspot/share/gc/g1/g1CollectedHeap.cpp	2020-09-04 13:56:57.789007593 +0200
@@ -1029,7 +1029,7 @@
 
   // Delete metaspaces for unloaded class loaders and clean up loader_data graph
   ClassLoaderDataGraph::purge();
-  DEBUG_ONLY(MetaspaceUtils::verify(false);)
+  DEBUG_ONLY(MetaspaceUtils::verify();)
 
   // Prepare heap for normal collections.
   assert(num_free_regions() == 0, "we should not have added any free regions");
