--- old/test/hotspot/jtreg/vmTestbase/metaspace/gc/MetaspaceBaseGC.java	2020-07-16 14:21:40.944369845 +0200
+++ new/test/hotspot/jtreg/vmTestbase/metaspace/gc/MetaspaceBaseGC.java	2020-07-16 14:21:40.608368567 +0200
@@ -101,7 +101,6 @@
 
     protected void configure(String args[]) {
         vmArgs.addAll(ManagementFactory.getRuntimeMXBean().getInputArguments());
-        useLargepages = PAGE_SIZE > 1_000_000 && !vmArgs.contains("-XX:-UseLargePagesInMetaspace");
 
         System.out.println(vmArgs);
 
