--- old/src/hotspot/share/jvmci/jvmciCompilerToVM.hpp	2020-07-16 14:20:45.340158334 +0200
+++ new/src/hotspot/share/jvmci/jvmciCompilerToVM.hpp	2020-07-16 14:20:45.012157086 +0200
@@ -25,6 +25,7 @@
 #define SHARE_JVMCI_JVMCICOMPILERTOVM_HPP
 
 #include "gc/shared/cardTable.hpp"
+#include "gc/shared/collectedHeap.hpp"
 #include "jvmci/jvmciExceptions.hpp"
 #include "runtime/javaCalls.hpp"
 #include "runtime/signature.hpp"
