--- old/src/hotspot/share/gc/parallel/psParallelCompact.cpp	2020-07-16 14:20:25.208081754 +0200
+++ new/src/hotspot/share/gc/parallel/psParallelCompact.cpp	2020-07-16 14:20:24.852080400 +0200
@@ -1060,7 +1060,7 @@
 
   // Delete metaspaces for unloaded class loaders and clean up loader_data graph
   ClassLoaderDataGraph::purge();
-  MetaspaceUtils::verify_metrics();
+  DEBUG_ONLY(MetaspaceUtils::verify(false);)
 
   heap->prune_scavengable_nmethods();
 
