--- old/src/hotspot/share/services/diagnosticCommand.cpp	2020-09-04 13:57:15.749122523 +0200
+++ new/src/hotspot/share/services/diagnosticCommand.cpp	2020-09-04 13:57:15.365120039 +0200
@@ -30,7 +30,7 @@
 #include "compiler/compileBroker.hpp"
 #include "compiler/directivesParser.hpp"
 #include "gc/shared/gcVMOperations.hpp"
-#include "memory/metaspace/metaspaceDCmd.hpp"
+#include "memory/metaspace/msDCmd.hpp"
 #include "memory/resourceArea.hpp"
 #include "memory/universe.hpp"
 #include "oops/objArrayOop.inline.hpp"
