--- old/test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/LargePages.java	2020-07-16 14:21:39.784365433 +0200
+++ new/test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/LargePages.java	2020-07-16 14:21:39.436364109 +0200
@@ -46,8 +46,8 @@
 
         SharedStringsUtils.dump(TestCommon.list("HelloString"),
             "SharedStringsBasic.txt", CDS_LOGGING,
-            "-XX:+UseLargePages", "-XX:+UseLargePagesInMetaspace");
+            "-XX:+UseLargePages");
         SharedStringsUtils.runWithArchive("HelloString",
-            "-XX:+UseLargePages", "-XX:+UseLargePagesInMetaspace");
+            "-XX:+UseLargePages");
     }
 }
