--- old/src/hotspot/share/memory/metaspace/metaspaceDCmd.hpp	2020-07-16 14:21:00.016214161 +0200
+++ new/src/hotspot/share/memory/metaspace/metaspaceDCmd.hpp	2020-07-16 14:20:59.660212805 +0200
@@ -1,6 +1,6 @@
 /*
- * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2018, SAP and/or its affiliates.
+ * Copyright (c) 2018, 2019 SAP and/or its affiliates.
+ * Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -38,7 +38,6 @@
   DCmdArgument<bool> _by_spacetype;
   DCmdArgument<bool> _by_chunktype;
   DCmdArgument<bool> _show_vslist;
-  DCmdArgument<bool> _show_vsmap;
   DCmdArgument<char*> _scale;
   DCmdArgument<bool> _show_classes;
 public:
