--- old/src/hotspot/share/memory/metaspace/metaspaceSizesSnapshot.cpp	2020-07-16 14:21:01.156218496 +0200
+++ new/src/hotspot/share/memory/metaspace/metaspaceSizesSnapshot.cpp	2020-07-16 14:21:00.808217173 +0200
@@ -26,6 +26,7 @@
 #include "precompiled.hpp"
 
 #include "memory/metaspace.hpp"
+#include "memory/metaspace.hpp"
 #include "memory/metaspace/metaspaceSizesSnapshot.hpp"
 
 namespace metaspace {
