--- old/src/hotspot/share/memory/metaspace/printMetaspaceInfoKlassClosure.cpp	2020-07-16 14:21:06.512238870 +0200
+++ new/src/hotspot/share/memory/metaspace/printMetaspaceInfoKlassClosure.cpp	2020-07-16 14:21:06.148237485 +0200
@@ -1,6 +1,6 @@
 /*
- * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2018, SAP and/or its affiliates.
+ * Copyright (c) 2018, 2019 SAP SE. All rights reserved.
+ * Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
