--- old/test/hotspot/jtreg/gc/TestSystemGC.java	2020-07-16 14:21:29.792327425 +0200
+++ new/test/hotspot/jtreg/gc/TestSystemGC.java	2020-07-16 14:21:29.416325994 +0200
@@ -44,8 +44,7 @@
  * @run main/othervm -XX:+UseG1GC gc.TestSystemGC
  * @run main/othervm -XX:+UseG1GC -XX:+ExplicitGCInvokesConcurrent gc.TestSystemGC
  * @run main/othervm -XX:+UseLargePages gc.TestSystemGC
- * @run main/othervm -XX:+UseLargePages -XX:+UseLargePagesInMetaspace gc.TestSystemGC
- */
+  */
 
 /*
  * @test TestSystemGCShenandoah
