--- old/test/hotspot/jtreg/runtime/CompressedOops/CompressedClassSpaceSize.java	2020-07-16 14:21:34.848346657 +0200
+++ new/test/hotspot/jtreg/runtime/CompressedOops/CompressedClassSpaceSize.java	2020-07-16 14:21:34.496345317 +0200
@@ -64,12 +64,14 @@
 
 
         // Make sure the minimum size is set correctly and printed
+        // (Note: ccs size shall be rounded up to the minimum size of 4m since metaspace reservations
+        //  are done in a 4m granularity. Note that this is **reserved** size and does not affect rss.
         pb = ProcessTools.createJavaProcessBuilder("-XX:+UnlockDiagnosticVMOptions",
                                                    "-XX:CompressedClassSpaceSize=1m",
                                                    "-Xlog:gc+metaspace=trace",
                                                    "-version");
         output = new OutputAnalyzer(pb.start());
-        output.shouldMatch("Compressed class space.*1048576")
+        output.shouldMatch("Compressed class space.*4194304")
               .shouldHaveExitValue(0);
 
 
