--- old/src/hotspot/share/prims/whitebox.cpp	2020-09-04 13:57:12.273100079 +0200
+++ new/src/hotspot/share/prims/whitebox.cpp	2020-09-04 13:57:11.897097657 +0200
@@ -46,7 +46,7 @@
 #include "memory/heapShared.inline.hpp"
 #include "memory/metaspaceShared.hpp"
 #include "memory/metadataFactory.hpp"
-#include "memory/metaspace/metaspace_test.hpp"
+#include "memory/metaspace/msTestHelpers.hpp"
 #include "memory/iterator.hpp"
 #include "memory/resourceArea.hpp"
 #include "memory/universe.hpp"
