--- old/src/hotspot/share/memory/universe.cpp	2020-09-04 13:57:11.069092326 +0200
+++ new/src/hotspot/share/memory/universe.cpp	2020-09-04 13:57:10.689089883 +0200
@@ -1163,7 +1163,7 @@
   }
   if (should_verify_subset(Verify_MetaspaceUtils)) {
     log_debug(gc, verify)("MetaspaceUtils");
-    DEBUG_ONLY(MetaspaceUtils::verify(true);)
+    DEBUG_ONLY(MetaspaceUtils::verify();)
   }
   if (should_verify_subset(Verify_JNIHandles)) {
     log_debug(gc, verify)("JNIHandles");
