--- old/src/hotspot/share/classfile/classLoaderDataGraph.hpp	2020-07-16 14:20:21.560067877 +0200
+++ new/src/hotspot/share/classfile/classLoaderDataGraph.hpp	2020-07-16 14:20:21.200066508 +0200
@@ -161,12 +161,4 @@
   static Klass* next_klass_in_cldg(Klass* klass);
 };
 
-class ClassLoaderDataGraphMetaspaceIterator : public StackObj {
-  ClassLoaderData* _data;
- public:
-  ClassLoaderDataGraphMetaspaceIterator();
-  ~ClassLoaderDataGraphMetaspaceIterator();
-  bool repeat() { return _data != NULL; }
-  ClassLoaderMetaspace* get_next();
-};
 #endif // SHARE_CLASSFILE_CLASSLOADERDATAGRAPH_HPP
