--- old/test/hotspot/jtreg/runtime/Metaspace/elastic/AllocationProfile.java	2020-09-04 13:57:32.777233827 +0200
+++ new/test/hotspot/jtreg/runtime/Metaspace/elastic/AllocationProfile.java	2020-09-04 13:57:32.521232138 +0200
@@ -48,7 +48,6 @@
         return r.nextInt((int)(maximumSingleAllocationSize - minimumSingleAllocationSize + 1)) + minimumSingleAllocationSize;
     }
 
-
     // Some standard profiles
     static final List<AllocationProfile> standardProfiles = new ArrayList<>();
 
