--- old/test/hotspot/jtreg/runtime/Metaspace/elastic/MetaspaceTestContext.java	2020-09-04 13:57:33.945241542 +0200
+++ new/test/hotspot/jtreg/runtime/Metaspace/elastic/MetaspaceTestContext.java	2020-09-04 13:57:33.677239772 +0200
@@ -23,7 +23,6 @@
     long numDeallocated;
     long allocationFailures;
 
-
     public MetaspaceTestContext(long commitLimit, long reserveLimit) {
         this.commitLimit = commitLimit;
         this.reserveLimit = reserveLimit;
@@ -126,7 +125,6 @@
      */
     public void checkStatistics() {
 
-
         // Note:
         // Estimating Used and Committed is fuzzy, and we only have limited information here
         // (we know the current state, but not the history, which determines fragmentation and
