--- old/test/hotspot/jtreg/runtime/Metaspace/elastic/MetaspaceTestOneArenaManyThreads.java	2020-09-04 13:57:34.505245247 +0200
+++ new/test/hotspot/jtreg/runtime/Metaspace/elastic/MetaspaceTestOneArenaManyThreads.java	2020-09-04 13:57:34.253243580 +0200
@@ -32,7 +32,6 @@
     // Several threads allocate from a single arena.
     // This mimicks several threads loading classes via the same class loader.
 
-
     public MetaspaceTestOneArenaManyThreads(MetaspaceTestContext context, long testAllocationCeiling, int numThreads, int seconds) {
         super(context, testAllocationCeiling, numThreads, seconds);
     }
