--- old/test/hotspot/jtreg/runtime/Metaspace/elastic/RandomAllocator.java	2020-09-04 13:57:35.677253004 +0200
+++ new/test/hotspot/jtreg/runtime/Metaspace/elastic/RandomAllocator.java	2020-09-04 13:57:35.425251336 +0200
@@ -102,7 +102,6 @@
         this.localRandom = new Random(RandomHelper.random().nextInt());
     }
 
-
     @Override
     public String toString() {
         return  arena.toString() + ", ticks=" + ticks;
