--- old/test/hotspot/jtreg/runtime/Metaspace/elastic/TestMetaspaceAllocation.java	2020-09-04 13:57:36.833260666 +0200
+++ new/test/hotspot/jtreg/runtime/Metaspace/elastic/TestMetaspaceAllocation.java	2020-09-04 13:57:36.573258942 +0200
@@ -25,6 +25,7 @@
 
 /*
  * @test id=debug
+ * @bug 8251158
  * @library /test/lib
  * @modules java.base/jdk.internal.misc
  *          java.management
@@ -42,6 +43,7 @@
 
 /*
  * @test id=ndebug
+ * @bug 8251158
  * @library /test/lib
  * @modules java.base/jdk.internal.misc
  *          java.management
