--- old/test/hotspot/jtreg/runtime/Metaspace/elastic/TestMetaspaceAllocationMT2.java	2020-09-04 13:57:38.049268736 +0200
+++ new/test/hotspot/jtreg/runtime/Metaspace/elastic/TestMetaspaceAllocationMT2.java	2020-09-04 13:57:37.729266613 +0200
@@ -74,7 +74,6 @@
  * @run main/othervm/timeout=400 -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI  -XX:MetaspaceReclaimPolicy=aggressive TestMetaspaceAllocationMT2
  */
 
-
 public class TestMetaspaceAllocationMT2 {
 
     public static void main(String[] args) throws Exception {
