--- old/test/hotspot/jtreg/runtime/cds/MaxMetaspaceSize.java	2020-07-16 14:21:38.576360837 +0200
+++ new/test/hotspot/jtreg/runtime/cds/MaxMetaspaceSize.java	2020-07-16 14:21:38.228359513 +0200
@@ -44,9 +44,8 @@
     processArgs.add("-Xshare:dump");
 
     if (Platform.is64bit()) {
-      processArgs.add("-XX:MaxMetaspaceSize=3m");
+      processArgs.add("-XX:MaxMetaspaceSize=8m");
       processArgs.add("-XX:CompressedClassSpaceSize=1m");
-      processArgs.add("-XX:InitialBootClassLoaderMetaspaceSize=1m");
     } else {
       processArgs.add("-XX:MaxMetaspaceSize=1m");
     }
