--- old/src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp	2020-07-16 14:20:36.980126534 +0200
+++ new/src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp	2020-07-16 14:20:36.616125148 +0200
@@ -72,7 +72,9 @@
 #include "gc/shenandoah/shenandoahJfrSupport.hpp"
 #endif
 
-#include "memory/metaspace.hpp"
+
+#include "memory/metaspace/classLoaderMetaspace.hpp"
+#include "memory/metaspace/metaspaceEnums.hpp"
 #include "oops/compressedOops.inline.hpp"
 #include "runtime/atomic.hpp"
 #include "runtime/globals.hpp"
@@ -2263,7 +2265,6 @@
   }
   // Resize and verify metaspace
   MetaspaceGC::compute_new_size();
-  MetaspaceUtils::verify_metrics();
 }
 
 // Weak roots are either pre-evacuated (final mark) or updated (final updaterefs),
