--- old/src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp	2020-07-16 14:20:38.184131113 +0200
+++ new/src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp	2020-07-16 14:20:37.856129866 +0200
@@ -35,6 +35,7 @@
 #include "gc/shenandoah/shenandoahPadding.hpp"
 #include "gc/shenandoah/shenandoahSharedVariables.hpp"
 #include "gc/shenandoah/shenandoahUnload.hpp"
+#include "memory/metaspace.hpp"
 #include "services/memoryManager.hpp"
 #include "utilities/globalDefinitions.hpp"
 
