--- old/test/hotspot/gtest/memory/test_virtualspace.cpp	2020-07-16 14:21:28.576322798 +0200
+++ new/test/hotspot/gtest/memory/test_virtualspace.cpp	2020-07-16 14:21:28.232321490 +0200
@@ -28,6 +28,7 @@
 #include "utilities/align.hpp"
 #include "unittest.hpp"
 
+
 namespace {
   class MemoryReleaser {
     ReservedSpace* const _rs;
@@ -337,3 +338,6 @@
   EXPECT_NO_FATAL_FAILURE(test_virtual_space_actual_committed_space(10 * M, 5 * M,  Commit));
   EXPECT_NO_FATAL_FAILURE(test_virtual_space_actual_committed_space(10 * M, 10 * M, Commit));
 }
+
+
+
