--- old/src/hotspot/share/gc/parallel/psParallelCompact.cpp	2020-09-04 13:56:59.457018158 +0200
+++ new/src/hotspot/share/gc/parallel/psParallelCompact.cpp	2020-09-04 13:56:59.089015824 +0200
@@ -1057,7 +1057,7 @@
 
   // Delete metaspaces for unloaded class loaders and clean up loader_data graph
   ClassLoaderDataGraph::purge();
-  DEBUG_ONLY(MetaspaceUtils::verify(false);)
+  DEBUG_ONLY(MetaspaceUtils::verify();)
 
   heap->prune_scavengable_nmethods();
 
