--- old/src/hotspot/share/runtime/globals.hpp	2020-07-16 14:21:21.500295882 +0200
+++ new/src/hotspot/share/runtime/globals.hpp	2020-07-16 14:21:21.148294544 +0200
@@ -897,7 +897,6 @@
           NOT_LP64(2200*K) LP64_ONLY(4*M),                                  \
           "(Deprecated) Initial size of the boot class loader data metaspace") \
           range(30*K, max_uintx/BytesPerWord)                               \
-          constraint(InitialBootClassLoaderMetaspaceSizeConstraintFunc, AfterErgo)\
                                                                             \
   product(bool, PrintHeapAtSIGBREAK, true,                                  \
           "Print heap layout in response to SIGBREAK")                      \
@@ -1548,6 +1547,18 @@
           "class pointers are used")                                        \
           range(1*M, 3*G)                                                   \
                                                                             \
+  experimental(uintx, ForceCompressedClassSpaceStartAddress, 0,             \
+          "Force class space start address to a given value.")              \
+                                                                            \
+  product(ccstr, MetaspaceReclaimStrategy, "balanced",                      \
+          "options: balanced, aggressive, none")                            \
+                                                                            \
+  develop(bool, MetaspaceGuardAllocations, false,                           \
+          "Metapace allocations are guarded.")                              \
+                                                                            \
+  develop(bool, MetaspaceHandleDeallocations, true,                         \
+	        "Switch off Metapace deallocation handling.")                     \
+																		                                        \
   manageable(uintx, MinHeapFreeRatio, 40,                                   \
           "The minimum percentage of heap free after GC to avoid expansion."\
           " For most GCs this applies to the old generation. In G1 and"     \
