--- old/src/hotspot/share/runtime/os.hpp	2020-07-16 14:21:22.680300372 +0200
+++ new/src/hotspot/share/runtime/os.hpp	2020-07-16 14:21:22.336299063 +0200
@@ -406,6 +406,8 @@
   static bool   can_execute_large_page_memory();
 
   // Check if pointer points to readable memory (by 4-byte read access)
+  // !! Unreliable before VM initialization! Use CanUseSafeFetch32() to test
+  //    if this function is reliable !!
   static bool    is_readable_pointer(const void* p);
   static bool    is_readable_range(const void* from, const void* to);
 
