--- old/src/hotspot/share/runtime/vmStructs.cpp	2020-07-16 14:21:25.124309668 +0200
+++ new/src/hotspot/share/runtime/vmStructs.cpp	2020-07-16 14:21:24.772308328 +0200
@@ -51,7 +51,6 @@
 #include "memory/allocation.hpp"
 #include "memory/allocation.inline.hpp"
 #include "memory/heap.hpp"
-#include "memory/metaspace/metablock.hpp"
 #include "memory/padded.hpp"
 #include "memory/referenceType.hpp"
 #include "memory/universe.hpp"
