--- old/src/hotspot/share/runtime/globals.hpp	2020-09-04 13:57:13.465107765 +0200
+++ new/src/hotspot/share/runtime/globals.hpp	2020-09-04 13:57:13.061105158 +0200
@@ -1560,8 +1560,8 @@
           "Metapace allocations are guarded.")                              \
                                                                             \
   develop(bool, MetaspaceHandleDeallocations, true,                         \
-	        "Switch off Metapace deallocation handling.")                     \
-																		                                        \
+          "Switch off Metapace deallocation handling.")                     \
+                                                                            \
   manageable(uintx, MinHeapFreeRatio, 40,                                   \
           "The minimum percentage of heap free after GC to avoid expansion."\
           " For most GCs this applies to the old generation. In G1 and"     \
