--- old/src/hotspot/share/runtime/vmOperations.cpp	2020-09-04 13:57:14.609115152 +0200
+++ new/src/hotspot/share/runtime/vmOperations.cpp	2020-09-04 13:57:14.233112722 +0200
@@ -33,7 +33,7 @@
 #include "logging/logStream.hpp"
 #include "logging/logConfiguration.hpp"
 #include "memory/heapInspection.hpp"
-#include "memory/metaspace/metaspaceReport.hpp"
+#include "memory/metaspace/msReport.hpp"
 #include "memory/resourceArea.hpp"
 #include "memory/universe.hpp"
 #include "oops/symbol.hpp"
