--- old/src/hotspot/share/memory/allocation.hpp	2020-07-16 14:20:46.528162853 +0200
+++ new/src/hotspot/share/memory/allocation.hpp	2020-07-16 14:20:46.180161529 +0200
@@ -141,6 +141,7 @@
   f(mtSafepoint,      "Safepoint")                                                   \
   f(mtSynchronizer,   "Synchronization")                                             \
   f(mtServiceability, "Serviceability")                                              \
+  f(mtMetaspace,      "Metaspace")                                                   \
   f(mtNone,           "Unknown")                                                     \
   //end
 
