--- old/src/hotspot/share/memory/filemap.cpp	2020-07-16 14:20:47.700167311 +0200
+++ new/src/hotspot/share/memory/filemap.cpp	2020-07-16 14:20:47.364166033 +0200
@@ -40,6 +40,7 @@
 #include "memory/heapShared.inline.hpp"
 #include "memory/iterator.inline.hpp"
 #include "memory/metadataFactory.hpp"
+#include "memory/metaspace.hpp"
 #include "memory/metaspaceClosure.hpp"
 #include "memory/metaspaceShared.hpp"
 #include "memory/oopFactory.hpp"
