--- old/src/hotspot/share/memory/metadataFactory.hpp	2020-07-16 14:20:48.872171769 +0200
+++ new/src/hotspot/share/memory/metadataFactory.hpp	2020-07-16 14:20:48.520170430 +0200
@@ -26,6 +26,8 @@
 #define SHARE_MEMORY_METADATAFACTORY_HPP
 
 #include "classfile/classLoaderData.hpp"
+#include "memory/metaspace/classLoaderMetaspace.hpp"
+#include "memory/metaspace/metaspaceEnums.hpp"
 #include "oops/array.hpp"
 #include "utilities/exceptions.hpp"
 #include "utilities/globalDefinitions.hpp"
