--- old/src/hotspot/share/memory/metaspaceClosure.hpp	2020-07-16 14:21:16.872278278 +0200
+++ new/src/hotspot/share/memory/metaspaceClosure.hpp	2020-07-16 14:21:16.524276954 +0200
@@ -106,7 +106,7 @@
   //         Symbol*     bar() { return (Symbol*)    _obj; }
   //
   // [2] All Array<T> dimensions are statically declared.
-  class Ref : public CHeapObj<mtInternal> {
+  class Ref : public CHeapObj<mtMetaspace> {
     Writability _writability;
     Ref* _next;
     NONCOPYABLE(Ref);
