--- old/src/hotspot/share/memory/metaspaceCounters.cpp	2020-07-16 14:21:17.976282478 +0200
+++ new/src/hotspot/share/memory/metaspaceCounters.cpp	2020-07-16 14:21:17.640281199 +0200
@@ -29,7 +29,7 @@
 #include "runtime/perfData.hpp"
 #include "utilities/exceptions.hpp"
 
-class MetaspacePerfCounters: public CHeapObj<mtInternal> {
+class MetaspacePerfCounters: public CHeapObj<mtMetaspace> {
   friend class VMStructs;
   PerfVariable*      _capacity;
   PerfVariable*      _used;
