--- old/src/hotspot/share/memory/metaspaceChunkFreeListSummary.hpp	2020-07-16 14:21:15.632273561 +0200
+++ new/src/hotspot/share/memory/metaspaceChunkFreeListSummary.hpp	2020-07-16 14:21:15.292272268 +0200
@@ -25,6 +25,7 @@
 #ifndef SHARE_MEMORY_METASPACECHUNKFREELISTSUMMARY_HPP
 #define SHARE_MEMORY_METASPACECHUNKFREELISTSUMMARY_HPP
 
+#include "utilities/globalDefinitions.hpp"
 
 class MetaspaceChunkFreeListSummary {
   size_t _num_specialized_chunks;
