--- old/src/hotspot/share/memory/universe.cpp	2020-07-16 14:21:20.332291439 +0200
+++ new/src/hotspot/share/memory/universe.cpp	2020-07-16 14:21:19.984290115 +0200
@@ -1100,7 +1100,7 @@
   }
   if (should_verify_subset(Verify_MetaspaceUtils)) {
     log_debug(gc, verify)("MetaspaceUtils");
-    MetaspaceUtils::verify_free_chunks();
+    DEBUG_ONLY(MetaspaceUtils::verify(true);)
   }
   if (should_verify_subset(Verify_JNIHandles)) {
     log_debug(gc, verify)("JNIHandles");
