--- old/src/hotspot/share/gc/shared/collectedHeap.cpp	2020-07-16 14:20:26.388086243 +0200
+++ new/src/hotspot/share/gc/shared/collectedHeap.cpp	2020-07-16 14:20:26.052084965 +0200
@@ -37,6 +37,7 @@
 #include "gc/shared/memAllocator.hpp"
 #include "logging/log.hpp"
 #include "memory/metaspace.hpp"
+#include "memory/metaspace/classLoaderMetaspace.hpp"
 #include "memory/resourceArea.hpp"
 #include "memory/universe.hpp"
 #include "oops/instanceMirrorKlass.hpp"
