--- old/src/hotspot/share/gc/shared/collectedHeap.hpp	2020-07-16 14:20:27.556090686 +0200
+++ new/src/hotspot/share/gc/shared/collectedHeap.hpp	2020-07-16 14:20:27.212089377 +0200
@@ -29,6 +29,7 @@
 #include "gc/shared/gcWhen.hpp"
 #include "gc/shared/verifyOption.hpp"
 #include "memory/allocation.hpp"
+#include "memory/metaspace.hpp"
 #include "memory/universe.hpp"
 #include "runtime/handles.hpp"
 #include "runtime/perfData.hpp"
