--- old/src/hotspot/share/gc/shared/gcTrace.cpp	2020-07-16 14:20:28.772095311 +0200
+++ new/src/hotspot/share/gc/shared/gcTrace.cpp	2020-07-16 14:20:28.428094002 +0200
@@ -30,6 +30,7 @@
 #include "gc/shared/gcTrace.hpp"
 #include "gc/shared/objectCountEventSender.hpp"
 #include "gc/shared/referenceProcessorStats.hpp"
+#include "memory/metaspace.hpp"
 #include "memory/heapInspection.hpp"
 #include "memory/resourceArea.hpp"
 #include "runtime/os.hpp"
