--- old/src/hotspot/share/gc/shared/gcVMOperations.cpp	2020-07-16 14:20:29.928099708 +0200
+++ new/src/hotspot/share/gc/shared/gcVMOperations.cpp	2020-07-16 14:20:29.580098385 +0200
@@ -32,6 +32,7 @@
 #include "gc/shared/genCollectedHeap.hpp"
 #include "interpreter/oopMapCache.hpp"
 #include "logging/log.hpp"
+#include "memory/metaspace/classLoaderMetaspace.hpp"
 #include "memory/oopFactory.hpp"
 #include "memory/universe.hpp"
 #include "runtime/handles.inline.hpp"
