--- old/src/hotspot/share/gc/shared/gcVMOperations.hpp	2020-07-16 14:20:31.064104029 +0200
+++ new/src/hotspot/share/gc/shared/gcVMOperations.hpp	2020-07-16 14:20:30.732102767 +0200
@@ -28,6 +28,7 @@
 #include "gc/shared/collectedHeap.hpp"
 #include "gc/shared/genCollectedHeap.hpp"
 #include "memory/heapInspection.hpp"
+#include "memory/metaspace.hpp"
 #include "prims/jvmtiExport.hpp"
 #include "runtime/handles.hpp"
 #include "runtime/jniHandles.hpp"
