--- old/src/hotspot/share/gc/shared/generationSpec.cpp	2020-07-16 14:20:33.464113159 +0200
+++ new/src/hotspot/share/gc/shared/generationSpec.cpp	2020-07-16 14:20:33.124111865 +0200
@@ -25,7 +25,6 @@
 #include "precompiled.hpp"
 #include "gc/shared/cardTableRS.hpp"
 #include "gc/shared/generationSpec.hpp"
-#include "memory/binaryTreeDictionary.hpp"
 #include "memory/filemap.hpp"
 #include "runtime/java.hpp"
 #include "utilities/macros.hpp"
