--- old/src/hotspot/share/gc/shared/jvmFlagConstraintsGC.hpp	2020-07-16 14:20:35.832122167 +0200
+++ new/src/hotspot/share/gc/shared/jvmFlagConstraintsGC.hpp	2020-07-16 14:20:35.476120813 +0200
@@ -56,7 +56,6 @@
 
 JVMFlag::Error MaxGCPauseMillisConstraintFunc(uintx value, bool verbose);
 JVMFlag::Error GCPauseIntervalMillisConstraintFunc(uintx value, bool verbose);
-JVMFlag::Error InitialBootClassLoaderMetaspaceSizeConstraintFunc(size_t value, bool verbose);
 JVMFlag::Error MinHeapSizeConstraintFunc(size_t value, bool verbose);
 JVMFlag::Error InitialHeapSizeConstraintFunc(size_t value, bool verbose);
 JVMFlag::Error MaxHeapSizeConstraintFunc(size_t value, bool verbose);
