--- old/test/hotspot/gtest/metaspace/test_metaspaceUtils.cpp	2020-09-04 13:57:28.725207139 +0200
+++ new/test/hotspot/gtest/metaspace/test_metaspaceUtils.cpp	2020-09-04 13:57:28.465205431 +0200
@@ -1,6 +1,6 @@
 /*
- * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2018, 2020 SAP SE. All rights reserved.
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020 SAP SE. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -27,7 +27,6 @@
 #include "memory/metaspace.hpp"
 #include "unittest.hpp"
 
-
 TEST_VM(MetaspaceUtils, reserved) {
   size_t reserved = MetaspaceUtils::reserved_bytes();
   EXPECT_GT(reserved, 0UL);
