--- old/test/hotspot/gtest/metaspace/test_is_metaspace_obj.cpp	2020-09-04 13:57:26.877195009 +0200
+++ new/test/hotspot/gtest/metaspace/test_is_metaspace_obj.cpp	2020-09-04 13:57:26.597193173 +0200
@@ -1,6 +1,6 @@
 /*
- * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2018, 2020 SAP SE. All rights reserved.
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020 SAP SE. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -27,7 +27,7 @@
 
 #include "memory/allocation.inline.hpp"
 #include "memory/metaspace.hpp"
-#include "memory/metaspace/virtualSpaceList.hpp"
+#include "memory/metaspace/msVirtualSpaceList.hpp"
 #include "runtime/mutex.hpp"
 #include "runtime/mutexLocker.hpp"
 #include "runtime/os.hpp"
@@ -35,8 +35,6 @@
 
 using namespace metaspace;
 
-
-
 // Test the cheerful multitude of metaspace-contains-functions.
 class MetaspaceIsMetaspaceObjTest {
   Mutex* _lock;
