--- old/src/hotspot/share/gc/g1/g1CollectedHeap.cpp	2020-07-16 14:20:23.940076932 +0200
+++ new/src/hotspot/share/gc/g1/g1CollectedHeap.cpp	2020-07-16 14:20:23.596075622 +0200
@@ -1029,7 +1029,7 @@
 
   // Delete metaspaces for unloaded class loaders and clean up loader_data graph
   ClassLoaderDataGraph::purge();
-  MetaspaceUtils::verify_metrics();
+  DEBUG_ONLY(MetaspaceUtils::verify(false);)
 
   // Prepare heap for normal collections.
   assert(num_free_regions() == 0, "we should not have added any free regions");
