--- old/test/hotspot/jtreg/gc/metaspace/CompressedClassSpaceSizeInJmapHeap.java	2020-07-16 14:21:32.348337147 +0200
+++ new/test/hotspot/jtreg/gc/metaspace/CompressedClassSpaceSizeInJmapHeap.java	2020-07-16 14:21:31.988335777 +0200
@@ -32,7 +32,7 @@
  * @library /test/lib
  * @modules java.base/jdk.internal.misc
  *          java.management
- * @run main/othervm -XX:+IgnoreUnrecognizedVMOptions -XX:CompressedClassSpaceSize=50m gc.metaspace.CompressedClassSpaceSizeInJmapHeap
+ * @run main/othervm -XX:+IgnoreUnrecognizedVMOptions -XX:CompressedClassSpaceSize=48m gc.metaspace.CompressedClassSpaceSizeInJmapHeap
  */
 
 import jdk.test.lib.JDKToolLauncher;
@@ -67,7 +67,7 @@
         run(pb);
 
         OutputAnalyzer output = new OutputAnalyzer(read(out));
-        output.shouldContain("CompressedClassSpaceSize = 52428800 (50.0MB)");
+        output.shouldContain("CompressedClassSpaceSize = 50331648 (48.0MB)");
         out.delete();
     }
 
