--- old/src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp	2020-09-04 13:57:01.801033043 +0200
+++ new/src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp	2020-09-04 13:57:01.421030627 +0200
@@ -2202,7 +2202,7 @@
   }
   // Resize and verify metaspace
   MetaspaceGC::compute_new_size();
-  DEBUG_ONLY(MetaspaceUtils::verify(false);)
+  DEBUG_ONLY(MetaspaceUtils::verify();)
 }
 
 // Weak roots are either pre-evacuated (final mark) or updated (final updaterefs),
