--- old/src/hotspot/share/memory/metaspace/metaspaceSizesSnapshot.cpp	2020-09-04 13:57:08.789077678 +0200
+++ new/src/hotspot/share/memory/metaspace/metaspaceSizesSnapshot.cpp	2020-09-04 13:57:08.397075165 +0200
@@ -25,9 +25,8 @@
 
 #include "precompiled.hpp"
 
-#include "memory/metaspace.hpp"
-#include "memory/metaspace.hpp"
 #include "memory/metaspace/metaspaceSizesSnapshot.hpp"
+#include "memory/metaspace.hpp"
 
 namespace metaspace {
 
