--- old/src/hotspot/share/memory/metaspace/metaspaceSizesSnapshot.hpp	2020-09-04 13:57:09.881084689 +0200
+++ new/src/hotspot/share/memory/metaspace/metaspaceSizesSnapshot.hpp	2020-09-04 13:57:09.505082275 +0200
@@ -26,6 +26,8 @@
 #ifndef SHARE_MEMORY_METASPACE_METASPACESIZESSNAPSHOT_HPP
 #define SHARE_MEMORY_METASPACE_METASPACESIZESSNAPSHOT_HPP
 
+#include "utilities/globalDefinitions.hpp"
+
 namespace metaspace {
 
 // Todo: clean up after jep387, see JDK-8251392
